### PR TITLE
Add transactions support to Db and use them when moving carrier

### PR DESCRIPTION
**Pull request by rmilecki**
```
Carrier: use transaction when moving carrier

It's required to keep database in consistent state if a later query
fails for whatever reason. As much as it's unlikely in this simple case
it's still a good practice and provides an example of using
transactions.
```
```
Db: add support for SQL transactions

Transactions should be used when executing multiple related queries.
They are required to keep database in consistent state in case of any
errors.

MySQL supports transactions only when using InnoDB however this code is
safe even for MyISAM. PHP "checks for transaction capabilities" for a
given driver only. So this code won't cause exception on MyISAM (it'll
just do nothing).

Ref: https://www.php.net/manual/en/pdo.transactions.php
```

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | All related database change queries should be executed in a single transaction. This way we make sure database won't get in inconsistent state if processing fails at some point.<br>This Pull Request adds transaction support to `Db` API and uses it when moving carrier. PrestaShop code should be updated incrementally to use transactions in all critical actions.
| Type?         | bug fix / improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Those changes are **inspired** by my problems with **orders** inconsistenty as reported in the [#17347](https://github.com/PrestaShop/PrestaShop/issues/17347). As this point I focused on moving carrier though (to start with something simple).
| How to test?  | Navigate to Carrier & make sure carriers can be moved up/down.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/17360)

### CODE REVIEW

**Comment by rmilecki** [comment]
> This is your first pull request on the PrestaShop project.

Dear @prestonBot (maintainer), it's my [4th pull request](https://github.com/PrestaShop/PrestaShop/pulls?q=author%3Armilecki), you may stop welcoming me now I think :wink:

File: controllers/admin/AdminOrdersController.php [hunk 1]
```diff
@@ -2125,6 +2125,8 @@ public function ajaxProcessAddProductOnOrder()
         // Total method
         $total_method = Cart::BOTH_WITHOUT_SHIPPING;
 
+        Db::getInstance()->beginTransaction();
```
Comment (PierreRambaud) [inline controllers/admin/AdminOrdersController.php:2128]: This kind of translations must be done only if innodb is used

Comment (rmilecki) [reply controllers/admin/AdminOrdersController.php:2128]: Good point! I'll fix that!
> If the underlying driver does not support transactions, a PDOException will be thrown (regardless of your error handling settings: this is always a serious error condition).

Reference: https://www.php.net/manual/en/pdo.transactions.php

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
We still support MyISAM

**Comment by Progi1984** [comment]
@PrestaShop/prestashop-core-developers What do you think about this PR ?

**Comment by Progi1984** [comment]
Hello @rmilecki, thanks for your contribution, would you mind to rebase your pull request for fixing conflicting file and fixing the check for InnoDB ? Thanks

**Comment by PierreRambaud** [comment]
Hi @rmilecki ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

File: controllers/admin/AdminOrdersController.php (hunk 1 above)
Comment (rmilecki) [reply controllers/admin/AdminOrdersController.php:2128]: @PierreRambaud: I've checked PHP documentation and it says it's actually **safe** to use transactions even with MyISAM. This code **won't** trigger an exception. I've tested it with PrestaShop `develop` branch running on MySQL 5.7.35 with all tables using MyISAM.

**Comment by rmilecki** [comment]
> Feel free to reopen or open another one if you think it's still relevant.

@PierreRambaud, @Progi1984: I can't find a way to reopen this pull request. Would you be so kind to reopen it for me, please? I've updated my branch.

**Comment by Progi1984** [comment]
@rmilecki I can't too. Could you recreate a new PR ?

**Comment by PierreRambaud** [comment]
@rmilecki the branch was force-pushed or recreated that's why we are not able to reopen the pr ^^

**Pull request closed without merging**